### ISSUE TITLE
Too many patches when saving refs?

### ISSUE THREAD

**Issue by mrchess** [codepunkt/mongoose-patch-history#6]
I have two schemas defined as follows:

I have two schemas defined as follows.

```javascript
const PersonSchema = new mongoose.Schema({
  organization: {
    type: mongoose.Schema.Types.ObjectId,
    ref: ORGANIZATION_MODEL_NAME,
    required: true,
  },
  name: {
    type: String,
    unique: true,
    required: true,
  }
})

const OrganizationSchema = new mongoose.Schema({
  name: {
    type: String,
    unique: true,
    required: true,
  }
})
```

I can create a person, no problem.

```javascript
Organization
  .create({ name: 'Home '})
  .then((org) => Person.create({ name: 'Bob', organization: organization._id })
  .then((person) => {
    // Now I have `person`
  });
```

However, if I try to change the org on the person:

`person.set({ organization: 'MY_NEW_ID' }).save();`

Patches looks like this:

```
DATE/OP/PATH/VALUE
2017-02-10T06:26:29.312Z  replace /organization/id/10 231 
2017-02-10T06:26:29.312Z  replace /organization/id/9  108 
2017-02-10T06:26:29.312Z  replace /organization/id/8  82  
2017-02-10T06:26:29.312Z  replace /organization/id/7  24  
2017-02-10T06:26:29.312Z  replace /organization/id/6  9 
2017-02-10T06:26:29.312Z  replace /organization/id/5  211 
2017-02-10T06:26:29.312Z  replace /organization/id/4  12  
2017-02-10T06:26:29.312Z  replace /organization/id/3  234 
2017-02-10T06:26:29.312Z  replace /organization/id/2  74
```

I was expecting more something like:

```
DATE/OP/PATH/VALUE
2017-02-10T06:26:29.312Z  replace /organization/id/2  MY_NEW_ID
```

Am I doing something wrong?

**Comment by mrchess** [comment]
Basically I don't understand how to save the Person -> Organization reference correctly.

**Comment by codepunkt** [comment]
@mrchess
1. Do you have a more complete example where i can also see the model connections being made and the patches being defined?
2. Before changing the ORG on the person, what do the person and organization collections/documents look like?

**Comment by mrchess** [comment]
Hmm, I tried a simple sandbox case and it seems to work. Perhaps it is the way I implemented it in my app. Pasting this here for reference. I'll try a few more cases.

```javascript
const mongoose = require('mongoose');
const patchHistory = require('mongoose-patch-history');
const assert = require('assert');
const Promise = require('bluebird');
mongoose.Promise = Promise;

// Organization Model
const OrganizationSchema = new mongoose.Schema({
  name: {
    type: String,
  },
});
const Organization = mongoose.model('Organization', OrganizationSchema);

// Person Model
const PersonSchema = new mongoose.Schema({
  organization: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Organization',
  },
  name: {
    type: String,
  },
});

PersonSchema.plugin(patchHistory.default, {
  mongoose,
  name: 'roomPatches',
});

const Person = mongoose.model('Person', PersonSchema);

//////////////
// The Test //
//////////////
const uri = 'mongodb://localhost:27017/mongoose_patch_history_test';

let workOrg;
let homeOrg;
let person;
mongoose.connect(uri)
  // Setup
  .then(() => Organization.create({ name: 'Home '}))
  .then((o) => (homeOrg = o))
  .then(() => Organization.create({ name: 'Work '}))
  .then((o) => (workOrg = o))
  .then(() => Person.create({ name: 'Bob', organization: homeOrg._id }))
  .then((p) => (person = p))
  // Test
  .then(() => person.patches.find({ ref: person.id }))
  .then((patches) => {
    console.log(JSON.stringify(patches,null,2));
    // [
    //   {
    //     "_id": "589d7124c2d0b1aa30dcfb14",
    //     "ref": "589d7124c2d0b1aa30dcfb13",
    //     "__v": 0,
    //     "ops": [
    //       {
    //         "op": "add",
    //         "path": "/name",
    //         "value": "Bob"
    //       },
    //       {
    //         "op": "add",
    //         "path": "/organization",
    //         "value": "589d7124c2d0b1aa30dcfb11"
    //       }
    //     ],
    //     "date": "2017-02-10T07:52:04.201Z"
    //   }
    // ]
  })
  // Switch the org
  .then(() => {
    person.organization = workOrg._id;
    return person.save();
  })
  .then(() => person.patches.find({ ref: person.id }))
  .then((patches) => {
    console.log(JSON.stringify(patches,null,2));
    // [
    //   {
    //     "_id": "589d7195101d63abfae70aee",
    //     "ref": "589d7195101d63abfae70aed",
    //     "__v": 0,
    //     "ops": [
    //       {
    //         "op": "add",
    //         "path": "/name",
    //         "value": "Bob"
    //       },
    //       {
    //         "op": "add",
    //         "path": "/organization",
    //         "value": "589d7195101d63abfae70aeb"
    //       }
    //     ],
    //     "date": "2017-02-10T07:53:57.053Z"
    //   },
    //   {
    //     "_id": "589d7195101d63abfae70aef",
    //     "ref": "589d7195101d63abfae70aed",
    //     "__v": 0,
    //     "ops": [
    //       {
    //         "op": "replace",
    //         "path": "/organization/id/11",
    //         "value": 236
    //       }
    //     ],
    //     "date": "2017-02-10T07:53:57.070Z"
    //   }
    // ]
  });
```

**Comment by mrchess** [comment]
Hmm, I tried a simple sandbox case and it seems to work. Perhaps it is the way I implemented it in my app. Pasting this here for reference. I'll try a few more cases.

The below code **WORKS AS EXPECTED**.

```javascript
const mongoose = require('mongoose');
const patchHistory = require('mongoose-patch-history');
const assert = require('assert');
const Promise = require('bluebird');
mongoose.Promise = Promise;

// Organization Model
const OrganizationSchema = new mongoose.Schema({
  name: {
    type: String,
  },
});
const Organization = mongoose.model('Organization', OrganizationSchema);

// Person Model
const PersonSchema = new mongoose.Schema({
  organization: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Organization',
  },
  name: {
    type: String,
  },
});

PersonSchema.plugin(patchHistory.default, {
  mongoose,
  name: 'roomPatches',
});

const Person = mongoose.model('Person', PersonSchema);

//////////////
// The Test //
//////////////
const uri = 'mongodb://localhost:27017/mongoose_patch_history_test';

let workOrg;
let homeOrg;
let person;
mongoose.connect(uri)
  // Setup
  .then(() => Organization.create({ name: 'Home '}))
  .then((o) => (homeOrg = o))
  .then(() => Organization.create({ name: 'Work '}))
  .then((o) => (workOrg = o))
  .then(() => Person.create({ name: 'Bob', organization: homeOrg._id }))
  .then((p) => (person = p))
  // Test
  .then(() => person.patches.find({ ref: person.id }))
  .then((patches) => {
    console.log(JSON.stringify(patches,null,2));
    // [
    //   {
    //     "_id": "589d7124c2d0b1aa30dcfb14",
    //     "ref": "589d7124c2d0b1aa30dcfb13",
    //     "__v": 0,
    //     "ops": [
    //       {
    //         "op": "add",
    //         "path": "/name",
    //         "value": "Bob"
    //       },
    //       {
    //         "op": "add",
    //         "path": "/organization",
    //         "value": "589d7124c2d0b1aa30dcfb11"
    //       }
    //     ],
    //     "date": "2017-02-10T07:52:04.201Z"
    //   }
    // ]
  })
  // Switch the org
  .then(() => {
    person.organization = workOrg._id;
    return person.save();
  })
  .then(() => person.patches.find({ ref: person.id }))
  .then((patches) => {
    console.log(JSON.stringify(patches,null,2));
    // [
    //   {
    //     "_id": "589d7195101d63abfae70aee",
    //     "ref": "589d7195101d63abfae70aed",
    //     "__v": 0,
    //     "ops": [
    //       {
    //         "op": "add",
    //         "path": "/name",
    //         "value": "Bob"
    //       },
    //       {
    //         "op": "add",
    //         "path": "/organization",
    //         "value": "589d7195101d63abfae70aeb"
    //       }
    //     ],
    //     "date": "2017-02-10T07:53:57.053Z"
    //   },
    //   {
    //     "_id": "589d7195101d63abfae70aef",
    //     "ref": "589d7195101d63abfae70aed",
    //     "__v": 0,
    //     "ops": [
    //       {
    //         "op": "replace",
    //         "path": "/organization/id/11",
    //         "value": 236
    //       }
    //     ],
    //     "date": "2017-02-10T07:53:57.070Z"
    //   }
    // ]
  });
```

Here is an example case. See how in the second patch the operation is kind of weird?

```
    //     "ops": [
    //       {
    //         "op": "replace",
    //         "path": "/organization/id/11",
    //         "value": 236
    //       }
```

**Comment by mrchess** [comment]
Quickly adding my version as well:

```javascript
  "dependencies": {
    "bluebird": "^3.4.7",
    "mongoose": "^4.8.1",
    "mongoose-patch-history": "^1.1.5"
  }
```

**Comment by codepunkt** [comment]
That seems weird indeed.

There is this line in the plugin:
`const ops = jsonpatch.compare(this.isNew ? {} : this._original, this.data())`
Can you see what is compared here? `this._original` vs `this.data()`

**Comment by mrchess** [comment]
Interesting it yields this:

```javascript
this._original // { name: 'Bob', organization: 589e95796dd6ed00f1d8c1ce } 
this.data() // { name: 'Bob', organization: 589e95796dd6ed00f1d8c1cf }
```

I'm trying to debug it.

**Comment by mrchess** [comment]
I think it is because they are mongoose ObjectId objects.

```javascript
console.log(typeof this._original.organization, typeof this.data().organization);
// object object
console.log(this._original.organization.constructor.name);
// ObjectID
```

**Comment by mrchess** [comment]
So yeah, it must be a problem with the `jsonpatch.compare` not being able to resolve these mongoose objects. Any suggestions? I may try to just change them to strings or something... the "ref" keys.

Have you yourself been using refs in your collections?

**Comment by mrchess** [comment]
Fwiw I think I'm just gonna change my:

```javascript
  organization: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Organization',
  },
```

to

```javascript
  organization: {
    type: String,
    ref: 'Organization',
  },
```

I don't really see a danger in doing this from a `mongoose` perspective. Please let me know if you see one.

**Comment by codepunkt** [comment]
Thanks for helping me out! 👍 

I need to think about this and play around with it when i find the time so i'm keeping this issue open. If ObjectId refs really can't be compared with `jsonpatch`, this should at least be documented in the readme!

**Comment by codepunkt** [comment]
@mrchess Actually, the comparison to generate JSON patches should be done on JSON data - which the object returned from `.data()` is not. I added in a `toJSON` method that converts BSON to JSON (and therefore ObjectId references to strings) before comparison.

Tested with your example, should work. Published on npm as v1.1.6! Enjoy! :)

**Comment by mrchess** [comment]
Nice change! 🎉
